packages/codecs-core/src/prefix-codec.ts

```diff
@@ -0,0 +1,118 @@
+import { assertByteArrayHasEnoughBytesForCodec } from './assertions';
+import {
+    Codec,
+    createDecoder,
+    createEncoder,
+    Decoder,
+    Encoder,
+    FixedSizeCodec,
+    FixedSizeDecoder,
+    FixedSizeEncoder,
+    getEncodedSize,
+    isFixedSize,
+    VariableSizeCodec,
+    VariableSizeDecoder,
+    VariableSizeEncoder,
+} from './codec';
+import { combineCodec } from './combine-codec';
+
+type NumberEncoder = Encoder<bigint | number> | Encoder<number>;
+type FixedSizeNumberEncoder<TSize extends number = number> =
+    | FixedSizeEncoder<bigint | number, TSize>
+    | FixedSizeEncoder<number, TSize>;
+type NumberDecoder = Decoder<bigint> | Decoder<number>;
+type FixedSizeNumberDecoder<TSize extends number = number> =
+    | FixedSizeDecoder<bigint, TSize>
+    | FixedSizeDecoder<number, TSize>;
+type NumberCodec = Codec<bigint | number, bigint> | Codec<number>;
+type FixedSizeNumberCodec<TSize extends number = number> =
+    | FixedSizeCodec<bigint | number, bigint, TSize>
+    | FixedSizeCodec<number, number, TSize>;
+
+/**
+ * Stores the size of the `encoder` in bytes as a prefix using the `prefix` encoder.
+ */
+export function prefixEncoder<TFrom>(
+    encoder: FixedSizeEncoder<TFrom>,
+    prefix: FixedSizeNumberEncoder,
+): FixedSizeEncoder<TFrom>;
+export function prefixEncoder<TFrom>(encoder: Encoder<TFrom>, prefix: NumberEncoder): VariableSizeEncoder<TFrom>;
+export function prefixEncoder<TFrom>(encoder: Encoder<TFrom>, prefix: NumberEncoder): Encoder<TFrom> {
+    const write = ((value, bytes, offset) => {
+        // Here we exceptionally use the `encode` function instead of the `write`
+        // function to contain the content of the encoder within its own bounds.
+        const encoderBytes = encoder.encode(value);
+        offset = prefix.write(encoderBytes.length, bytes, offset);
+        bytes.set(encoderBytes, offset);
```

Comment:
We have to do _something_ to prevent this scenario, right?

```ts
const { decode, encode } = prefixCodec(getUtf8Encoder(), getU8Encoder());
const bytes = encode('string longer [...] than 256 characters');
bytes >= 257; // true
const str = decode(bytes); // string longer [...] than 256 ch
// And now the buffer after that point is corrupt
```

Reply:
I'll had a test to be explicit about it but if the u8 codec isn't able to store the string size, it will throw an error.

It's the responsibility of the size codec to determine if it can indeed encode/decode the provided size because you could technically provide more complicated number codecs such as the `shortU16`.

Reply:
Ah, excellent.

Reply:
I added a test for this.